src/plus/gk/serverConnection.ts

```diff
@@ -169,28 +192,180 @@ export class ServerConnection implements Disposable {
 					url = `${url}?${options.query}`;
 				}
 			}
-			// TODO@eamodio handle common response errors
 
-			return this.fetch(
+			const rsp = await this.fetch(
 				url,
 				{
 					...init,
 					headers: headers as HeadersInit,
 				},
 				options,
 			);
+			if (!rsp.ok) {
+				await this.handleUnsuccessfulResponse(rsp, scope);
+			} else {
+				this.resetRequestExceptionCount();
+			}
+			return rsp;
 		} catch (ex) {
-			Logger.error(ex, scope);
+			this.handleRequestError('gitkraken', ex, scope);
 			throw ex;
 		}
 	}
 
+	private buildProviderRequestRateLimitError(token: string | undefined, ex: RequestError) {
+		let resetAt: number | undefined;
+
+		const reset = ex.response?.headers?.['x-ratelimit-reset'];
+		if (reset != null) {
+			resetAt = parseInt(reset, 10);
+			if (Number.isNaN(resetAt)) {
+				resetAt = undefined;
+			}
+		}
+		return new ProviderRequestRateLimitError(ex, token, resetAt);
+	}
+
+	private async handleUnsuccessfulResponse(rsp: Response, scope: LogScope | undefined): Promise<void> {
+		// Too Many Requests
+		if (rsp.status == 429) {
+			this.trackRequestException();
+			return;
+		}
+
+		// Forbidden
+		if (rsp.status == 403) {
+			if (rsp.statusText.includes('rate limit')) {
+				this.trackRequestException();
+			}
+			return;
+		}
+
+		// Internal Server Error
+		if (rsp.status == 500) {
+			this.trackRequestException();
+			void showIntegrationRequestFailed500WarningMessage(
+				'GitKraken failed to respond and might be experiencing issues. Please visit the [GitKraken status page](https://cloud.gitkrakenstatus.com) for more information.',
+			);
+			return;
+		}
+
+		const content = await rsp.text();
+
+		// Bad Gateway
+		if (rsp.status == 502) {
+			Logger.error(`GitKraken request failed: ${content} (${rsp.statusText})`, scope);
+			if (content.includes('timeout')) {
+				this.trackRequestException();
+				void showIntegrationRequestTimedOutWarningMessage('GitKraken');
+			}
+			return;
+		}
+
+		// Service Unavailable
+		if (rsp.status == 503) {
+			Logger.error(`GitKraken request failed: ${content} (${rsp.statusText})`, scope);
+			this.trackRequestException();
+			void showIntegrationRequestFailed500WarningMessage(
+				'GitKraken failed to respond and might be experiencing issues. Please visit the [GitKraken status page](https://cloud.gitkrakenstatus.com) for more information.',
+			);
+			return;
+		}
+
+		if (rsp.status >= 400 && rsp.status < 500) {
+			return;
+		}
+
+		if (Logger.isDebugging) {
+			void window.showErrorMessage(`GitKraken request failed: ${content} (${rsp.statusText})`);
+		}
+	}
+
+	private handleRequestError(
+		token: string | undefined,
+		ex: RequestError | (Error & { name: 'AbortError' }),
+		scope: LogScope | undefined,
+	): void {
+		if (ex instanceof CancellationError) throw ex;
+		if (ex.name === 'AbortError') throw new CancellationError(ex);
+
+		switch (ex.status) {
+			case 404: // Not found
+				throw new ProviderRequestNotFoundError(ex);
```

Comment:
The pattern "ProviderRequest....Error" that these are named after, refers to hosting providers or issue providers, like GitHub, GitLab, etc. Either we can name these "GitKrakenRequest...Error" or make the existing ones more generic like "Request...Error" and use that model for all of these. "ProviderRequest..." sounds integration-specific, which is not the case here.

Reply:
Done.
I've made it more generalized, because for now they don't contain anything provider specific and can be used for both services. We'll be able to introduce specific errors later, when/if they are truly needed.

https://github.com/gitkraken/vscode-gitlens/commit/1006ba3fc9059a8c2d295cf1ec4f029f4831a60f